tests/unit/schema/wrappers/test_routing.py

```diff
@@ -0,0 +1,127 @@
+        # This may raise exception (which we show to clients).
+        request_field_value = _get_field(request, routing_param.field)
+        if routing_param.path_template:
+            routing_param_regex = routing_param.to_regex()
+            regex_match = routing_param_regex.match(request_field_value)
+            if regex_match:
+                header_params[routing_param.key] = regex_match.group(
+                    routing_param.key)
+        else:  # No need to match
+            header_params[routing_param.key] = request_field_value
+    return header_params
+
+
+@pytest.mark.parametrize("req, expected", [
+    (RoutingTestRequest(app_profile_id="foo.123"),
+     {"app_profile_id": "foo.123"}),
+    (RoutingTestRequest(app_profile_id="projects/100"),
+     {"app_profile_id": "projects/100"}),
+    (RoutingTestRequest(app_profile_id=""), {"app_profile_id": ""}),
+])
+def test_routing_rule_resolve_simple_extraction(req, expected):
+    rule = wrappers.RoutingRule(
+        [wrappers.RoutingParameter("app_profile_id", "")]
+    )
+    assert resolve(rule, req) == expected
+
+
+@pytest.mark.parametrize("req, expected", [
+    (RoutingTestRequest(app_profile_id="foo.123"), {"routing_id": "foo.123"}),
+    (RoutingTestRequest(app_profile_id="projects/100"),
+     {"routing_id": "projects/100"}),
+    (RoutingTestRequest(app_profile_id=""), {"routing_id": ""}),
+])
+def test_routing_rule_resolve_rename_extraction(req, expected):
+    rule = wrappers.RoutingRule(
+        [wrappers.RoutingParameter("app_profile_id", "{routing_id=**}")]
+    )
+    assert resolve(rule, req) == expected
+
+
+@pytest.mark.parametrize("req, expected", [
+    (RoutingTestRequest(table_name="projects/100/instances/200"),
+     {"table_name": "projects/100/instances/200"}),
+    (RoutingTestRequest(table_name="projects/100/instances/200/whatever"),
+     {"table_name": "projects/100/instances/200/whatever"}),
+    (RoutingTestRequest(table_name="foo"), {}),
+])
+def test_routing_rule_resolve_field_match(req, expected):
+    rule = wrappers.RoutingRule(
+        [
+            wrappers.RoutingParameter(
+                "table_name", "{table_name=projects/*/instances/*/**}"),
+            wrappers.RoutingParameter(
+                "table_name", "{table_name=regions/*/zones/*/**}"),
+        ]
+    )
+    assert resolve(rule, req) == expected
+
+
+@pytest.mark.parametrize("routing_parameters, req, expected", [
+    ([wrappers.RoutingParameter("table_name", "{project_id=projects/*}/instances/*/**")],
+     RoutingTestRequest(table_name="projects/100/instances/200/tables/300"),
+     {"project_id": "projects/100"}),
+    ([wrappers.RoutingParameter("table_name", "{project_id=projects/*}/instances/*/**"),
+      wrappers.RoutingParameter(
+          "table_name", "projects/*/{instance_id=instances/*}/**"),
+      ],
+     RoutingTestRequest(table_name="projects/100/instances/200/tables/300"),
+     {"project_id": "projects/100", "instance_id": "instances/200"})
```

Comment:
This is a little hard to eyeball parse. Can you reformat it using black?

Reply:
I ran `black`, but it looks like style_check CI test fails. So, looks like the file needs to be formatted using `find gapic tests -name "*.py" -not -path 'tests/**/goldens/*' | xargs autopep8 --in-place`